### PR TITLE
To docker

### PR DESCRIPTION
**Pull request by kuhnen**

### CODE REVIEW

**Comment by parisni** [comment]
hey. While the glue binding is ok (listing databases, tables and so on), reading from s3 is stuck with those warning (same apply when reading s3 parquet path directly without glue):

```
21/08/10 17:42:45 WARN package: Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.sql.debug.maxToStringFields'.
21/08/10 17:42:45 WARN SessionState: METASTORE_FILTER_HOOK will be ignored, since hive.security.authorization.manager is set to instance of HiveAuthorizerFactory.
21/08/10 17:43:11 WARN MetricsConfig: Cannot locate configuration: tried hadoop-metrics2-s3a-file-system.properties,hadoop-metrics2.properties
```
Is it expected ? (I also tested to build spark with hadoop-cloud https://spark.apache.org/docs/latest/cloud-integration.html)

**Comment by kuhnen** [comment]
@parisni these warnings are ok.  I never had problems.

**Comment by parisni** [comment]
I had trouble with ssl on the s3 buckets. I added `--conf spark.hadoop.fs.s3a.connection.ssl.enabled=false` as a workaround.

The warnings get removed by copying the `hadoop-metrics2.properties` file present in the hadoop folder into spark/conf/

File: Dockerfile [hunk 1]
```diff
@@ -25,26 +21,71 @@ RUN echo "deb http://ftp.us.debian.org/debian sid main" >> /etc/apt/sources.list
...
+WORKDIR /opt
+RUN git clone https://github.com/apache/spark.git spark_clone
+##cd spark_clone
+WORKDIR /opt/spark_clone
+
+RUN git checkout "tags/v${SPARK_VERSION}" -b "v${SPARK_VERSION}"
+#RUN ./dev/make-distribution.sh --name spark-patched --pip -Pkubernetes -Phive -Phive-thriftserver -Phadoop-provided -Dhadoop.version="${HADOOP_VERSION}"
+RUN ./dev/make-distribution.sh --name spark-patched --pip -Phive -Phive-thriftserver -Phadoop-provided -Dhadoop.version="${HADOOP_VERSION}"
+
+COPY conf/* ./dist/conf
```
Comment (parisni) [inline Dockerfile:75]: add a missing `/` at the end of the folder (otherwize breaks on my docker versioin)
COPY conf/* ./dist/conf/

File: Dockerfile [hunk 2]
```diff
@@ -25,26 +21,71 @@ RUN echo "deb http://ftp.us.debian.org/debian sid main" >> /etc/apt/sources.list
...
+
+#BUILD SPARK
+WORKDIR /opt
+RUN git clone https://github.com/apache/spark.git spark_clone
+##cd spark_clone
+WORKDIR /opt/spark_clone
+
+RUN git checkout "tags/v${SPARK_VERSION}" -b "v${SPARK_VERSION}"
+#RUN ./dev/make-distribution.sh --name spark-patched --pip -Pkubernetes -Phive -Phive-thriftserver -Phadoop-provided -Dhadoop.version="${HADOOP_VERSION}"
+RUN ./dev/make-distribution.sh --name spark-patched --pip -Phive -Phive-thriftserver -Phadoop-provided -Dhadoop.version="${HADOOP_VERSION}"
```
Comment (parisni) [inline Dockerfile:73]: what about adding hadoop-cloud  cf https://spark.apache.org/docs/latest/cloud-integration.html

File: conf/spark-defaults.conf [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+spark.hadoop.fs.gs.impl com.google.cloud.hadoop.fs.gcs.GoogleHadoopFileSystem
+spark.hadoop.fs.AbstractFileSystem.gs.impl com.google.cloud.hadoop.fs.gcs.GoogleHadoopFS
+
+spark.hadoop.mapreduce.fileoutputcommitter.algorithm.version        2
+spark.hadoop.mapreduce.fileoutputcommitter.cleanup-failures.ignored true
+
+spark.hadoop.parquet.enable.summary-metadata false
+spark.sql.parquet.mergeSchema                false
+spark.sql.parquet.filterPushdown             true
+spark.sql.hive.metastorePartitionPruning     true
```
Comment (parisni) [inline conf/spark-defaults.conf:18]: maybe some optimization https://spark.apache.org/docs/latest/cloud-integration.html#committing-work-into-cloud-storage-safely-and-fast and  https://spark.apache.org/docs/latest/cloud-integration.html#parquet-io-settings ?

File: Dockerfile [hunk 4]
```diff
@@ -3,17 +3,13 @@ FROM python:3.7-slim-buster
 
 # Build options
 ARG hive_version=2.3.7
-ARG spark_version=3.0.0
+ARG spark_version=3.1.2
```
Comment (parisni) [inline Dockerfile:6]: spark 3.1.x has an issue with glue (date parition) https://github.com/awslabs/aws-glue-data-catalog-client-for-apache-hive-metastore/issues/45
3.0.2 is ok

**Comment by kuhnen** [comment]
@parisni thanks a lot for your comments :).
I will merge this to master and please feel free to create a pull request with your ideas 👍 .
Really nice 👍

**Pull request merged**